File: src/main/java/managers/QuestTemplateManager.java
package managers;

import client.player.quest.QuestTemplate;
import util.packet.PacketReader;

import java.util.HashMap;
import java.util.Map;

public class QuestTemplateManager extends AbstractManager {

    private static final Map<Integer, QuestTemplate> templates = new HashMap<>();

    public static QuestTemplate getQuest(int id) {
        QuestTemplate template = templates.get(id);
        if (template == null) {
            template = new QuestTemplate(id);
            loadQuestData(template);
            templates.put(id, template);
        }
        return template;
    }

    private static void loadQuestData(QuestTemplate template) {
        PacketReader r = getData("wz/Quest/" + template.getId() + ".mao");

        if (r != null) {
            int sflags = r.readInteger();
            int eflags = r.readInteger();

            template.getStartingRequirements().decode(sflags, r);
            template.getEndingRequirements().decode(eflags, r);

            //System.out.println("Finished initializing quest: " + template.getId());
        }
    }
}

File: src/main/java/managers/QuestTemplateManager.kt
package managers

import client.player.quest.QuestTemplate

object QuestTemplateManager : AbstractManager() {

    private val templates: MutableMap<Int, QuestTemplate> = HashMap()

    fun getQuest(id: Int): QuestTemplate? {
        var template = templates[id]
        if (template == null) {
            template = QuestTemplate(id)
            if (!loadQuestData(template)) return null
            templates[id] = template
        }
        return template
    }

    private fun loadQuestData(template: QuestTemplate): Boolean {
        val r = getData("wz/Quest/" + template.id + ".mao") ?: return false

        val sflags = r.readInteger()
        val eflags = r.readInteger()
        template.startingRequirements.decode(sflags, r)
        template.endingRequirements.decode(eflags, r)

        //System.out.println("Finished initializing quest: " + template.getId());
        return true
    }
}